shared/src/uniswap_pool.rs

```diff
@@ -19,10 +20,72 @@ pub trait PoolFetching: Send + Sync {
     async fn fetch(&self, token_pairs: HashSet<TokenPair>) -> Vec<Pool>;
 }
 
-#[derive(Clone)]
+#[derive(Clone, Hash)]
 pub struct Pool {
     pub tokens: TokenPair,
     pub reserves: (u128, u128),
+    pub fee: Rational,
+}
+
+impl Pool {
+    pub fn uniswap(tokens: TokenPair, reserves: (u128, u128)) -> Self {
+        Self {
+            tokens,
+            reserves,
+            fee: Rational::new(3, 1000),
+        }
+    }
+
+    /// Given an input amount and token, returns the maximum output amount and address of the other asset.
+    pub fn get_amount_out(&self, token_in: H160, amount_in: U256) -> (U256, H160) {
+        // https://github.com/Uniswap/uniswap-v2-periphery/blob/master/contracts/libraries/UniswapV2Library.sol#L43
+        let (reserve_in, reserve_out, token_out) = if token_in == self.tokens.get().0 {
+            (
+                U256::from(self.reserves.0),
+                U256::from(self.reserves.1),
+                self.tokens.get().1,
+            )
+        } else {
+            assert!(token_in == self.tokens.get().1, "Token not part of pool");
+            (
+                U256::from(self.reserves.1),
+                U256::from(self.reserves.0),
+                self.tokens.get().0,
+            )
+        };
+
+        let amount_in_with_fee = amount_in * (self.fee.denom() - self.fee.numer());
```

Comment:
I think this and the other method are missing some overflow checks like in https://github.com/gnosis/gp-v2-services/blob/c7b301a4fdfe1f92b691a5e5bf64af4ecee31428/solver/src/uniswap.rs (form an unmerged PR) .
Might also be nice to have amount_out and amount_in from that file with the way the arguments don't include any H160. That way those methods can't panic and testing is a bit nicer and we have the match on the H160 separately from the math.

Reply:
👍 on the missing overflow check. 

For the publicly facing interface, I find it nice to give callers the opportunity to get an amount without having to query the right reserve amount themselves from the outside. Would you suggest to split this into two methods then?

Reply:
Right you can still keep the current public method but have the inner in addition (maybe private).